### DISCUSSION TITLE
Running scans using a specific template version

### DISCUSSION THREAD

**Discussion by pdelteil** [projectdiscovery/nuclei#1125] [Q&A] [answered]
**Is your feature request related to a problem? Please describe.**
I'm often having difficulties running a scan using only the latest templates due to the use of axiom (deleted instances after every scan). 

**Describe the solution you'd like**
It would be useful to run a scan using as input the version of the templates release. 

Example:  ``nuclei -t ~/nuclei-templates -vt '8.5.9' ``

**Comment by ehsandeep** [comment]
@pdelteil to support features like this we need to have a copy of all versions on the machine and we only store the latest version of the template release, so this might be not possible to implement, as far as the use case is concerned, I can suggest using `nuclei -nt` on a new fleet of machines, as nuclei will auto-update latest templates as soon as you execute nuclei after template release creation.

**Comment by forgedhallpass** [comment]
@pdelteil 

You can clone a specific version (without history) into a directory of your choosing (e.g. `/tmp`): 
`git clone --depth=1 --branch v8.5.8 https://github.com/projectdiscovery/nuclei-templates.git` and then you can use `nuclei -update-directory /tmp/nuclei-templates-8.5.8 -no-update-templates`. You can easily create a shell script for this scenario, which would automatically do this for you, just don't forget to set the template directory back to your default one after you've finished

**Comment by pdelteil** [comment]
Thanks. I didn't think about that alternative. I will give it a try.

**Comment by pdelteil** [comment]
So, it works with this function:

`runScanTemplateVersion v8.5.8 urls.txt` 
```
runScanTemplateVersion()
{
    if [ -z "$1" ] | [ -z "$2" ]
    then
      echo "Use ${FUNCNAME[0]} version urls.txt"
      return 1;
    fi
    #git repo
    gitURL="https://github.com/projectdiscovery/nuclei-templates.git"
    branch="$1"
    file="$2"
    folder="/tmp/nuclei-templates-$branch"

    if [ -d "$folder" ] 
    then
        echo "Directory $folder exists. Skipping git clone" 
    else
        git clone --depth=1 --branch $branch $gitURL $folder
    fi

    nuclei -update-directory $folder -no-update-templates -l $file
}

```

The only problem is that running nuclei gives the following message (which is misleading):

![Screenshot from 2021-10-20 17-37-44](https://user-images.githubusercontent.com/20244863/138168463-dae13db8-291f-4bc7-8b86-03d6395a02e0.png)

**Comment by pdelteil** [comment]
Well, after testing it is not what I really intended to do, since every branch (or version) contains all previous branches/versions. I need to find a way to get only the new templates from version to version. 

I think this would work: 

```
git clone version1 folder1
git clone version2 folder2

#find not duplicated templates  
#copy the into /tmp/new-templates 
nuclei -update-directory  /tmp/new-templates  -no-update-templates -l $file 
```

**Reply by forgedhallpass** [reply]
Could you explain the motivation behind this? What is the added benefit of running the newly added templates only between versions? If you've ran your scans with the latest version of the templates and there is a new release, you have an option to run the newly added ones only.

**Comment by pdelteil** [comment]
I'm sorry for not answering it before. 

The idea is that sometimes I'm interested in running some new (and hot) templates, but not all (latest release). I could run them by tag or id manually, but it would take some manual work and time.

**Comment by forgedhallpass** [comment]
You could extend your script with something like the following:
```bash
git diff v8.8.3...v8.8.4 --name-only | grep ".yaml$"
```

This would return only the paths of the templates that were added or changed in the `v8.8.4` release. You can then iterate the results and execute new scans.

In https://github.com/projectdiscovery/nuclei/issues/1452, we'll probably add support for loading template paths from a file, meaning, you won't even need the iteration part anymore.

**Reply by forgedhallpass** [reply]
With the new `-template-url` flag, you could theoretically also host the resulting file via a `python3 -m http.server`, but that might be a little overkill 😄 (the `remote-template-domain` would also need to be changed)

**Reply by forgedhallpass** [reply]
#1452 was implemented, you should be able to reference the desired template paths from a file. Theoretically you should have everything in place to achieve what you wanted :)

**Reply by pdelteil** [reply]
> #1452 was implemented, you should be able to reference the desired template paths from a file. Theoretically you should have everything in place to achieve what you wanted :)

Thanks a lot. I will give it a try.

**Comment by forgedhallpass** [comment, marked as answer]
@pdelteil we've decided to add direct support for this feature:

[Option to run only new template by given version](https://github.com/projectdiscovery/nuclei/issues/2227)

You can try it out on the `dev` branch. It will be present in the next release :)

**Comment by pdelteil** [comment]
Thanks @forgedhallpass. 

I will give it a try!
